src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -3996,6 +4053,997 @@ End Class
+        }
+    }
+
+    static void Main()
+    {
+        M(new C(""a"", ""b""));
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+
+            Assert.Equal(
+                "void C.Deconstruct(out System.String? X, out System.String Y)",
+                comp.GetMember("C.Deconstruct").ToTestDisplayString(includeNonNullable: false));
+        }
+
+        [Fact]
+        public void Deconstruct_Conversion_03()
+        {
+            var source = @"
+using System;
+
+class Base { }
+class Derived : Base { }
+
+record C(Derived X, Base Y)
+{
+    public Base X { get; init; }
+    public Derived Y { get; init; }
+
+    static void M(C c)
+    {
+        switch (c)
+        {
+            case C(Derived x, Base y):
+                Console.Write(x);
+                Console.Write(y);
+                break;
+        }
+    }
+
+    static void Main()
+    {
+        M(new C(new Derived(), new Base()));
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (7,18): error CS8866: Record member 'C.X' must be a readable instance property of type 'Derived' to match positional parameter 'X'.
+                // record C(Derived X, Base Y)
+                Diagnostic(ErrorCode.ERR_BadRecordMemberForPositionalParameter, "X").WithArguments("C.X", "Derived", "X").WithLocation(7, 18),
+                // (7,26): error CS8866: Record member 'C.Y' must be a readable instance property of type 'Base' to match positional parameter 'Y'.
+                // record C(Derived X, Base Y)
+                Diagnostic(ErrorCode.ERR_BadRecordMemberForPositionalParameter, "Y").WithArguments("C.Y", "Base", "Y").WithLocation(7, 26));
+
+            Assert.Equal(
+                "void C.Deconstruct(out Derived X, out Base Y)",
+                comp.GetMember("C.Deconstruct").ToTestDisplayString(includeNonNullable: false));
+        }
+
+        [Fact]
+        public void Deconstruct_Empty_WithParameterList()
+        {
+            var source = @"
+record C()
+{
+    static void M(C c)
+    {
+        switch (c)
+        {
+            case C():
+                break;
+        }
+    }
+
+    static void Main()
+    {
+        M(new C());
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (2,9): error CS8850: A positional record must have both a 'data' modifier and non-empty parameter list
+                // record C()
+                Diagnostic(ErrorCode.ERR_BadRecordDeclaration, "()").WithLocation(2, 9));
+
+            Assert.Equal(
+                "void C.Deconstruct()",
+                comp.GetMember("C.Deconstruct").ToTestDisplayString(includeNonNullable: false));
+        }
+
+        [Fact]
+        public void Deconstruct_UserDefined()
+        {
+            var source =
+@"using System;
+
+record B(int X, int Y)
+{
+    public void Deconstruct(out int X, out int Y)
```

Comment:
Possible scenario to test: base type declares a public `Deconstruct` method (maybe virtual or maybe not) with the same signature as the `Deconstruct` that a derived type would synthesize, except for the return type.

```
record B(int X)
{
    public int Deconstruct(out int a, out int b) => throw null;
}
record C(int X, int Y) : B(X)
{
}
```

Reply:
Added the tests